### PR TITLE
IDENTITY-6629: OpenID Connect Hybrid Flow Implementation.

### PR DESCRIPTION
**Pull request by hasinidilanka**

### CODE REVIEW

**Comment by hasinidilanka** [comment]
@mefarazath @madurangasiriwardena @pulasthi7

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 1]
```diff
@@ -132,7 +132,9 @@
     private Map<String, String> supportedGrantTypeValidatorNames = new HashMap<>();
     private Map<String, Class<? extends OAuthValidator<HttpServletRequest>>> supportedGrantTypeValidators;
     private Map<String, String> supportedResponseTypeClassNames = new HashMap<>();
+    private Map<String, String> defaultSupporetdResponseTypeClassNames = new HashMap<>();
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: typo @ 'supported'

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 2]
```diff
@@ -666,6 +668,34 @@ public boolean isRefreshTokenRenewalEnabled() {
         return supportedResponseTypes;
     }
 
+    public Map<String, ResponseTypeHandler> getDefaultSupportedResponseTypes() {
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: Add doc comments describing what this method does

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 3]
```diff
@@ -666,6 +668,34 @@ public boolean isRefreshTokenRenewalEnabled() {
...
+        if (defaultSupportedResponseTypes == null) {
+            synchronized (this) {
+                if (defaultSupportedResponseTypes == null) {
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: you have initialized this value with an empty map. So, is there any possibility of this being null? (Did you intend an empty check instead? )

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 4]
```diff
@@ -666,6 +668,34 @@ public boolean isRefreshTokenRenewalEnabled() {
...
+                    Map<String, ResponseTypeHandler> defaultSupportedResponseTypesTemp = new Hashtable<>();
+                    for (Map.Entry<String, String> entry : defaultSupporetdResponseTypeClassNames.entrySet()) {
+                        ResponseTypeHandler responseTypeHandler = null;
+                        try {
+                            responseTypeHandler = (ResponseTypeHandler) Class.forName(entry.getValue()).newInstance();
+                            responseTypeHandler.init();
+                        } catch (InstantiationException e) {
+                            log.error("Error instantiating " + entry.getValue(), e);
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: Do we need separate catch clauses?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 5]
```diff
@@ -1651,6 +1684,37 @@ private void parseSupportedResponseTypesConfig(OMElement oauthConfigElem) {
         }
     }
 
+    private void parseDefaultSupportedResponseTypeConfig(List<String> supportedResponseTypes) {
+        if (!supportedResponseTypes.contains(ResponseType.CODE.toString())) {
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: Can you explain the logic here?
why are we checking !supportedResponseTypes.contains(ResponseType.CODE.toString() condition?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 6]
```diff
@@ -1651,6 +1684,37 @@ private void parseSupportedResponseTypesConfig(OMElement oauthConfigElem) {
...
+        if (!supportedResponseTypes.contains(ResponseType.TOKEN.toString())) {
+            for (String responseType : supportedResponseTypes) {
+                if (responseType.contains(ResponseType.TOKEN.toString())) {
+                    defaultSupporetdResponseTypeClassNames.put(ResponseType.TOKEN.toString(),
+                            "org.wso2.carbon.identity.oauth2.authz.handlers.TokenResponseTypeHandler");
+                    return;
+                }
+            }
+        }
+        if (!supportedResponseTypes.contains("id_token")) {
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: use a constant for "id_token"

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/AbstractResponseTypeHandler.java [hunk 7]
```diff
@@ -164,4 +171,36 @@ public boolean isAuthorizedClient(OAuthAuthzReqMessageContext authzReqMsgCtx) th
...
+        OAuthAppDO oAuthAppDO = null;
+        try {
+            oAuthAppDO = OAuth2Util.getAppInformationByClientId(consumerKey);
+        } catch (InvalidOAuthClientException e) {
+            throw new IdentityOAuth2Exception("Error while retrieving app information for clientId: " + consumerKey, e);
+        }
+        return oAuthAppDO;
+    }
+
+    public static ResponseTypeHandler getInitializedHandler(String responseType) {
```
Comment (hasinidilanka) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/AbstractResponseTypeHandler.java]: remove

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/CodeResponseTypeHandler.java [hunk 8]
```diff
@@ -40,30 +36,34 @@
...
-        OAuthAppDO oAuthAppDO = AppInfoCache.getInstance().getValueFromCache(authorizationReqDTO.getConsumerKey());
-        if (oAuthAppDO == null) {
-            try {
-                oAuthAppDO = new OAuthAppDAO().getAppInformation(authorizationReqDTO.getConsumerKey());
-            } catch (InvalidOAuthClientException e) {
-                throw new IdentityOAuth2Exception("Invalid consumer application. Failed to issue Grant token.", e);
-            }
-            AppInfoCache.getInstance().addToCache(oAuthAppDO.getOauthConsumerKey(), oAuthAppDO);
-        }
+    public void buildCode(OAuthAuthzReqMessageContext oauthAuthzMsgCtx, OAuth2AuthorizeRespDTO respDTO)
```
Comment (hasinidilanka) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/CodeResponseTypeHandler.java]: return respDTO. GenerateAuthorizationCode

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/CodeResponseTypeHandler.java [hunk 9]
```diff
@@ -40,30 +36,34 @@
...
+            throws IdentityOAuth2Exception {
 
+        OAuth2AuthorizeReqDTO authorizationReqDTO = oauthAuthzMsgCtx.getAuthorizationReqDTO();
+        String authorizationCode;
+        String codeId;
```
Comment (hasinidilanka) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/CodeResponseTypeHandler.java]: initialize here

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/IDTokenResponseTypeHandler.java [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+/**
+ * IDTokenResponseTypeHandler class creates an id_token when the response type is equal to id_token.
+ */
+public class IDTokenResponseTypeHandler extends AbstractResponseTypeHandler {
+
+    @Override
+    public OAuth2AuthorizeRespDTO issue(OAuthAuthzReqMessageContext oauthAuthzMsgCtx) throws IdentityOAuth2Exception {
+        ResponseTypeHandlerUtil.triggerPreListeners(oauthAuthzMsgCtx);
+        OAuth2AuthorizeRespDTO respDTO = initResponse(oauthAuthzMsgCtx);
```
Comment (hasinidilanka) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/IDTokenResponseTypeHandler.java]: move to buildIDTokenRespDTO

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java [hunk 11]
```diff
@@ -116,11 +116,11 @@ public Application updateApplication(ApplicationUpdateRequest updateRequest, Str
                 appDTO.setApplicationName(updateRequest.getClientName());
             }
             if (!updateRequest.getGrantTypes().isEmpty()) {
-                String grantType = StringUtils.join(updateRequest.getGrantTypes(), " ");
```
Comment (sagara-gunathunga) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java]: Shall we have a constant here for " " ?

File: components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMServiceTest.java [hunk 12]
```diff
@@ -511,9 +511,12 @@ public void registerApplicationTestWithFailedToUpdateSP(List<String> redirectUri
 
         oAuthConsumerApp.setGrantTypes(grantType);
         oAuthConsumerApp.setOAuthVersion(OAUTH_VERSION);
+        oAuthConsumerApp.setOauthConsumerKey("dummyConsumerKey");
```
Comment (sagara-gunathunga) [inline components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMServiceTest.java]: Shall we have a constant here ?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/util/ResponseTypeHandlerUtil.java [hunk 13]
```diff
@@ -0,0 +1,617 @@
...
+                if (log.isDebugEnabled()) {
+                    log.debug("Oauth post token issue listener is triggered.");
+                }
+            } catch (IdentityOAuth2Exception e) {
+                log.error("Oauth post token issue listener ", e);
+            }
+        }
+    }
+
+    public static AccessTokenDO generateAccessToken(OAuthAuthzReqMessageContext oauthAuthzMsgCtx, boolean cacheEnabled,
```
Comment (sagara-gunathunga) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/util/ResponseTypeHandlerUtil.java]: This is a very long method, you  need to break this into number of meaningful  methods.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/util/ResponseTypeHandlerUtil.java [hunk 14]
```diff
@@ -0,0 +1,617 @@
...
+                    log.debug("Access Token was added to OAuthCache for cache key : " + accessTokenCacheKey
+                            .getCacheKeyString());
+                }
+            }
+
+        }
+        return  tokenDO;
+    }
+
+    public static AuthzCodeDO generateAuthorizationCode(OAuthAuthzReqMessageContext oauthAuthzMsgCtx, boolean cacheEnabled,
```
Comment (sagara-gunathunga) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/util/ResponseTypeHandlerUtil.java]: This method also too long

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/jwks/JwksEndpoint.java [hunk 15]
```diff
@@ -17,186 +17,126 @@
...
-            } else {
-                buffer.append(ENCODE_MAP[p0]);
-                buffer.append(ENCODE_MAP[p1 | p2]);
-                if (padding) {
-                    buffer.append("==");
-                }
-            }
+    private String getTenantDomain() {
+        Object tenantObj = IdentityUtil.threadLocalProperties.get().get(OAuthConstants.TENANT_NAME_FROM_CONTEXT);
+        if (tenantObj != null && StringUtils.isNotBlank((String) tenantObj)) {
```
Comment (madumalt) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/jwks/JwksEndpoint.java]: No need to do check null here. StringUtils.isNotBlank does it  for us. :)

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java (hunk 11 above)
Comment (mefarazath) [reply components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java]: I guess its already added in line 119

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 16]
```diff
@@ -1777,12 +1780,14 @@ private void parseSupportedResponseTypesConfig(OMElement oauthConfigElem) {
             // if this element is not present, assume the default case.
             log.warn("\'SupportedResponseTypes\' element not configured in identity.xml. " +
                     "Therefore instantiating default response type handlers");
-
             Map<String, String> defaultResponseTypes = new HashMap<>(4);
             defaultResponseTypes.put(ResponseType.CODE.toString(), "org.wso2.carbon.identity.oauth2.authz.handlers.CodeResponseTypeHandler");
-            defaultResponseTypes.put(ResponseType.TOKEN.toString(), "org.wso2.carbon.identity.oauth2.authz.handlers.TokenResponseTypeHandler");
-            defaultResponseTypes.put("id_token", "org.wso2.carbon.identity.oauth2.authz.handlers.TokenResponseTypeHandler");
-            defaultResponseTypes.put("id_token token", "org.wso2.carbon.identity.oauth2.authz.handlers.TokenResponseTypeHandler");
+            defaultResponseTypes.put(ResponseType.TOKEN.toString(), "org.wso2.carbon.identity.oauth2.authz.handlers.AccessTokenResponseTypeHandler");
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: For consistency sake, shall we add "token" as a constant to OAuthConstants

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 17]
```diff
@@ -1777,12 +1780,14 @@ private void parseSupportedResponseTypesConfig(OMElement oauthConfigElem) {
             // if this element is not present, assume the default case.
             log.warn("\'SupportedResponseTypes\' element not configured in identity.xml. " +
                     "Therefore instantiating default response type handlers");
-
             Map<String, String> defaultResponseTypes = new HashMap<>(4);
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: We are initializing a map of 4 elements and adding more than four elements. Why define the size at all?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/AccessTokenResponseTypeHandler.java [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+ * AccessTokenResponseTypeHandler class generates an access token.
+ */
+public class AccessTokenResponseTypeHandler extends AbstractResponseTypeHandler {
+
+    private static Log log = LogFactory.getLog(AccessTokenResponseTypeHandler.class);
+
+    @Override
+    public OAuth2AuthorizeRespDTO issue(OAuthAuthzReqMessageContext oauthAuthzMsgCtx)
+            throws IdentityOAuth2Exception {
+        //Starting to trigger pre listeners.
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/AccessTokenResponseTypeHandler.java]: Leave a space after // and end comments with a full stop.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/HybridResponseTypeHandler.java [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception;
+import org.wso2.carbon.identity.oauth2.authz.OAuthAuthzReqMessageContext;
+import org.wso2.carbon.identity.oauth2.authz.handlers.util.ResponseTypeHandlerUtil;
+import org.wso2.carbon.identity.oauth2.dto.OAuth2AuthorizeReqDTO;
+import org.wso2.carbon.identity.oauth2.dto.OAuth2AuthorizeRespDTO;
+import org.wso2.carbon.identity.oauth2.model.AccessTokenDO;
+import org.wso2.carbon.identity.oauth2.model.AuthzCodeDO;
+
+/**
+ * HybridResponseTypeHandler class will handle the response when it is equal to
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/HybridResponseTypeHandler.java]: it --> response_type parameter

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/HybridResponseTypeHandler.java [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+
+        //Generating authorization code and generating response for authorization code flow.
+        if (StringUtils.contains(responseType, ResponseType.CODE.toString())) {
+            AuthzCodeDO authzCodeDO = ResponseTypeHandlerUtil.generateAuthorizationCode(oauthAuthzMsgCtx, cacheEnabled,
+                    oauthIssuerImpl);
+            ResponseTypeHandlerUtil.buildAuthorizationCodeResponseDTO(respDTO, authzCodeDO);
+        }
+
+        //Generating access token and generating response for access token flow.
+        if (StringUtils.contains(responseType, ResponseType.TOKEN.toString()) &&
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/HybridResponseTypeHandler.java]: Shall we move 
`StringUtils.contains(responseType, ResponseType.TOKEN.toString()) && !responseType.equalsIgnoreCase(OAuthConstants.CODE_IDTOKEN)`

to a private method that reflects its purpose. Something like isAcceessTokenIssued()

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/HybridResponseTypeHandler.java [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+                !responseType.equalsIgnoreCase(OAuthConstants.CODE_IDTOKEN)) {
+            AccessTokenDO accessTokenDO = ResponseTypeHandlerUtil.generateAccessToken(oauthAuthzMsgCtx, cacheEnabled,
+                    oauthIssuerImpl);
+            ResponseTypeHandlerUtil.buildAccessTokenResponseDTO(respDTO, accessTokenDO);
+        }
+
+        //Generating id_token and generating response for id_token flow.
+        if (StringUtils.contains(responseType, OAuthConstants.ID_TOKEN)) {
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/HybridResponseTypeHandler.java]: Shall we move `StringUtils.contains(responseType, OAuthConstants.ID_TOKEN)` to a private method which reflects the meaning of why we do this check?
Something like isIDTokenGenerated() or isIDTokenApplicableForResponseType()

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/util/ResponseTypeHandlerUtil.java (hunk 13 above)
Comment (hasinidilanka) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/util/ResponseTypeHandlerUtil.java]: This is an already existing method and I have moved it to the util class.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/util/ResponseTypeHandlerUtil.java (hunk 14 above)
Comment (hasinidilanka) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/util/ResponseTypeHandlerUtil.java]: This is an already existing method and I have moved it to the util class.

**Comment by isharak** [comment]
@hasinidilanka 
Can you send this PR to a feature branch?

**Comment by emswbandara** [comment]
Closing this since PR is merged with cd64dfa61f4065cdb800bc8ed6e0026c811801dc

**Pull request closed without merging**